driver-core/src/main/com/mongodb/internal/connection/SaslAuthenticator.java

```diff
@@ -131,6 +112,50 @@ private void throwIfSaslClientIsNull(final SaslClient saslClient) {
         }
     }
 
+    private BsonDocument getNextSaslResponse(final SaslClient saslClient, final InternalConnection connection) {
+        BsonDocument response = getSpeculativeAuthenticateResponse();
+        if (response != null) {
+            return response;
+        }
+
+        try {
+            byte[] serverResponse = saslClient.hasInitialResponse() ? saslClient.evaluateChallenge(new byte[0]) : null;
+            return sendSaslStart(serverResponse, connection);
+        } catch (Exception e) {
+            throw wrapException(e);
+        }
+    }
+
+    private void getNextSaslResponseAsync(final SaslClient saslClient, final InternalConnection connection,
+                                          final SingleResultCallback<Void> callback) {
+        BsonDocument authenticateResponse = getSpeculativeAuthenticateResponse();
```

Comment:
nit: Naming - can we name as per the sync method? `authenticateResponse` = `response`.

Reply:
Done